external-crates/move/crates/move-vm-runtime/src/tracing2/tracer.rs

```diff
@@ -0,0 +1,1812 @@
+                }
+                RuntimeLocation::Local(fidx, lidx) => Location::Local(*fidx, *lidx),
+                RuntimeLocation::Indexed(loc, idx) => {
+                    Location::Indexed(Box::new(loc.into_trace_location()), *idx)
+                }
+                RuntimeLocation::Global(id) => Location::Global(*id),
+            }
+        }
+
+        fn into_runtime_location(loc: Location) -> Self {
+            match loc {
+                Location::Local(fidx, lidx) => RuntimeLocation::Local(fidx, lidx),
+                Location::Indexed(loc, idx) => RuntimeLocation::Indexed(
+                    Box::new(RuntimeLocation::into_runtime_location(*loc)),
+                    idx,
+                ),
+                Location::Global(id) => RuntimeLocation::Global(id),
+            }
+        }
+    }
+
+    impl LocalType {
+        fn into_rooted_type(self) -> Option<RootedType> {
+            let ref_type = match self.ref_type {
+                ReferenceType::Value => None,
+                ReferenceType::Empty { .. } => panic!("Empty reference type"),
+                ReferenceType::Filled { ref_type, location } => Some((ref_type, location)),
+            };
+            Some(RootedType {
+                layout: self.layout?,
+                ref_type,
+            })
+        }
+    }
+
+    impl RootedType {
+        fn into_local_type(self) -> LocalType {
+            let ref_type = match self.ref_type {
+                None => ReferenceType::Value,
+                Some((ref_type, location)) => ReferenceType::Filled { ref_type, location },
+            };
+            LocalType {
+                layout: Some(self.layout),
+                ref_type,
+            }
+        }
+    }
+
+    impl<'a> VMTracer<'a> {
+        /// Unwrap or emit an error event to the trace
+        fn unwrap_or_emit_err(&mut self, x: Option<()>) {
+            x.unwrap_or_else(|| {
+                self.trace.effect(EF::ExecutionError(
+                    "!! TRACING ERROR !! Events below this may be incorrect.".to_string(),
+                ));
+            });
+        }
+
+        /// Get the current locals type and reference state(s)
+        fn current_frame_locals(&self) -> Option<&[LocalType]> {
+            Some(
+                self.active_frames
+                    .last_key_value()?
+                    .1
+                    .locals_types
+                    .as_slice(),
+            )
+        }
+
+        /// Return the current frame identifier. This is trace index of the frame and is used to
+        /// identify reference locations rooted higher up the call stack.
+        fn current_frame_identifier(&self) -> TraceIndex {
+            *self.active_frames.last_key_value().unwrap().0
+        }
+
+        /// Given the trace index for a frame, return the index of the frame in the call stack.
+        fn trace_index_to_frame_index(&self, idx: TraceIndex) -> Option<usize> {
```

Comment:
Can you generally expect trace indices to be monotonic (increasing) in the active frames list, and therefore can you make this a bisect instead of a linear scan? Not that the performance matters, but the fact that it was a linear scan made me think about this

Reply:
You should be able to. I can take a look :)

Reply:
Fixed so this does a binary search for it :)